### DISCUSSION TITLE
Prisma generated date type / Next.js and dates

### DISCUSSION THREAD

**Discussion by marceloverdijk** [prisma/prisma#4428] [Q&A] [answered]
I'm still experimenting with Next.js together with Prisma using TypeScript.
The first impression was good but now I came across some undesired behaviour.

As the Next.js page props are serialized as a JSON object it does not support dates.
When trying to pass a Prisma object with a date field in results in:
`'object' ("[object Date]") cannot be serialized as JSON. Please only return JSON serializable data types.`

Now I was looking at the Primas [example](https://github.com/prisma/prisma-examples/blob/latest/typescript/rest-nextjs-api-routes-auth/pages/p/%5Bid%5D.tsx) using Next.js and I noticed that dates seem to be excluded, on purpose?
E.g the `Post` model has no date and the date fields in the `User` model are not queried...

Is this done on purpose? And is there small 'workaround' how to pass models with date using Next.js.

**Comment by ryands17** [comment, marked as answer]
Hey @marceloverdijk 👋 

You're most probable facing an issue that Next.js has where it doesn't serialize the `Date` type.

I would suggest reading the below discussions/issues and the workaround is to stringify the date first and then set them in props.

https://github.com/vercel/next.js/issues/11993
https://github.com/vercel/next.js/discussions/11498

**Reply by marceloverdijk** [reply]
Yes I read these already and it seems a `JSON.parse(JSON.stringify(data, replacer))` is the best option.; with the `replacer` doing a string conversion of dates.

I understand the rationale behind it but if feels a bit cumbersome and was thinking if there would be nicer workaround.
For now I will stick with the `replacer`. Thx!

**Reply by marceloverdijk** [reply]
The problem with stringifying the date in the props is that the returned type is not the actual type anymore. So - at least when using TypeScript - the Prisma models cannot be passed through the components which is a pity.

It means I have to create different types for each Prisma model but with dates as strings unfortunately. This is not an issue with Prisma though, but a limitation how Next.js passes data...

I wonder if I can do something smart as a workaround...

**Reply by marceloverdijk** [reply]
Following up on this, in the Prisma docs I read:
https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference/#remarks-3

> DateTime
> See also: DateTime mappings in the nativeTypes Preview (2.11.0)

> Remarks
> Prisma Client returns all DateTime as ISO 8601-formatted strings.

I wonder what the last part means, as the Prisma client generator transforms a field like `dateOfBirth DateTime?` to `dateOfBirth: Date | null`.

Is it possible to get a ISO-8601 string from Prisma?

**Reply by marceloverdijk** [reply]
Maybe the docs are wrong about `Prisma Client returns all DateTime as ISO 8601-formatted strings.` as it seems this was the behavior before but changed in https://github.com/prisma/prisma-client-js/issues/260

**Reply by ryands17** [reply]
@marceloverdijk Yeah, currently you can only use ISO strings while inserting data. When you fetch the data it will always be mapped to the `Date` object.

**Reply by marceloverdijk** [reply]
Yes, I thought so.

Probably I have to create "JSON" DTO's for all my Prisma models and map queried objects to the DTO's. 
These DTO instances can then be safely passed to Next.js.

Another option is to have a computed/virtual dateAsString property in the Prisma models and a way to exclude the actual date from the select. For both a github issues but not implemented yet. Although I must admit that would feel a bit hackish as well...
I really wonder which patterns other users are implementing to solve this. I mean every app has dates....

**Reply by ryands17** [reply]
I would suggest taking a look at how [Blitz](https://github.com/blitz-js/blitz) does it. It uses a custom library made by them called [superjson](https://github.com/blitz-js/superjson) and the perform the serialization automatically at the app layer. In this case as you're using vanilla Next.js, you would need to add the plugin and do not need to implement the serialization yourself which is awesome :)

**Reply by marceloverdijk** [reply]
Thx @ryands17 for pointing me to superjson. It's awesome and does exactly what I need!

**Reply by slugmandrew** [reply]
It's mid-2022 and I have just added `superjson` to my Next.js app... it was super easy (using the babel plugin), seems to work great and now I can use my prisma types in my Next.js pages. 

Slightly surprised how long this has been an issue though.

**Reply by alex-zwingli** [reply]
Configuring `superjson` was a breeze and instantly solved this issue! Thank you so much for posting this @ryands17 🙏

**Reply by nbibler** [reply]
Just a heads up that in 2025, with Next.js 15, the seeming abandonment of `next-superjson-plugin` and `babel-plugin-superjson-next`, there is not currently a clear, simple, "just install this" path to addressing this issue, anymore.

**Comment by yahya-aghdam** [comment]
I converted the obj to json string then passed it to client. For other functions I converted the json to obj again in client. 

I hope Next.js add this ability later!

**Reply by Talor-A** [reply]
You might be interested in blitzjs.com - specifically the library it uses under the hood for parsing JSON on the client: [superjson](https://github.com/blitz-js/superjson)

Edit: looks like blitz already got a shoutout in another thread :)

**Comment by nodexchange** [comment]
For anyone 'not-too-keen' on blitz reverse engineering exercise. 

I did some basic remapping like this: 
```
export async function getServerSideProps(context) {
  let users = await prisma.user.findMany({
    where: { paid: false },
    select: {
      email: true,
      paid: true,
      name: true,
      createdAt: true,
    },
  });
  users.map((item) => {
    item.createdAt = item.createdAt.toString();
  });

  return {
    props: {
      users,
      session: await getSession(context),
    },
  };
}
```

**Reply by bradgreens** [reply]
> For anyone 'not-too-keen' on blitz reverse engineering exercise.
> 
> I did some basic remapping like this:

You can avoid manually identifying properties like `createdAt` by doing the following. In my case I had other properties which prisma is returning as `Date`.

```
  let listings = await prisma.listing.findMany({
    where: {
      userId: appUser.id,
    },
  });

  listings.forEach((listing) => {
    Object.entries(listing).forEach(([key, prop]) => {
      if (prop instanceof Date) {
        listing[key] = prop.toString();
      }
    });
  });
```

`map` isn't actually needed in this example since we're just mutating the original objects. But it'd probably be better to move this into an immutable helper function with `map`.

**Reply by im-paul-mak33** [reply]
Thanks for the help here you both, worked charm.

**Reply by im-paul-mak33** [reply]
```
    const listings = await prisma.listing.findMany({
      where: query,
      orderBy: {
        createdAt: 'desc',
      },
    });

    const safeListings = listings.map((listing) => ({
      ...listing,
      createdAt: listing.createdAt.toUTCString(),
    }));
```

Here is how I have done it, fetched the blog details and spefically stringified the creation date.
output - Sun, 10 Sep 2023 10:13:55 GMT

can find out more about the stringification of date on this link - [here](https://www.prisma.io/dataguide/mongodb/working-with-dates)

**Comment by robchava** [comment]
The following code works for me, and preserves the types. This is from SO, but don't have the source:
```typescript
type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> & Pick<U, Extract<keyof U, keyof T>> extends infer O
  ? {
      [K in keyof O]: O[K];
    }
  : never;

export const recursivelySerializeDatesInObject = <T extends { updatedAt?: Date; createdAt: Date }>(
  obj: T
): Overwrite<T, { updatedAt: string; createdAt: string }> => {
  const retObj: any = { ...obj };

  if (obj.createdAt) {
    retObj.createdAt = obj.createdAt.toISOString();
  }

  if (obj.updatedAt) {
    retObj.updatedAt = obj.updatedAt.toISOString();
  }

  Object.keys(retObj).forEach((key) => {
    if (retObj[key] && typeof retObj[key] === `object`) {
      retObj[key] = recursivelySerializeDatesInObject(retObj[key]);
    }
  });

  return retObj as any;
};
```

**Comment by najibghadri** [comment]
Anyone with a solution for next 14? It still doesn't work... Is that superjson solution deprecated for next.js? Should I just use JSON.serialize/parse? :D 

    "next-superjson-plugin": "0.5.10", or  "^0.6.3",
    "next": "^14.1.1",
    "superjson": "^1.13.3",

Tried other versions too, nothing worked...     Also tried next 13
